src/cargo/core/resolver/mod.rs

```diff
@@ -298,109 +244,213 @@ fn activate(mut cx: Box<Context>,
+        id: id,
+    }))
 }
 
-/// Activates the dependencies for a package, one by one in turn.
-///
-/// This function will attempt to activate all possible candidates for each
-/// dependency of the package specified by `parent`. The `deps` iterator
-/// provided is an iterator over all dependencies where each element yielded
-/// informs us what the candidates are for the dependency in question.
-///
-/// The `platform` argument is the target platform that the dependencies are
-/// being activated for.
+#[derive(Clone)]
+struct RcVecIter<Elem> {
+    vec: Rc<Vec<Elem>>,
+    next_index: usize,
+}
+
+impl<Elem> RcVecIter<Elem> {
+    fn new(vec: Vec<Elem>) -> RcVecIter<Elem> {
+        RcVecIter {
+            vec: Rc::new(vec),
+            next_index: 0
+        }
+    }
+    fn next(&mut self) -> Option<(usize, &Elem)> {
+        match self.vec.get(self.next_index) {
+            None => None,
+            Some(val) => {
+                let index = self.next_index;
+                self.next_index += 1;
+                Some((index, val))
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+struct DepsFrame {
+    parent: Rc<Summary>,
+    remaining_siblings: RcVecIter<DepInfo>,
+    platform: Option<String>,
+    id: PackageId,
+}
+type RemainingDeps = Vec<DepsFrame>;
+
+struct BacktrackFrame {
+    context_backup: Context,
+    deps_backup: RemainingDeps,
+    remaining_candidates: RcVecIter<Rc<Summary>>,
+    // For building an activation error:
+    parent: Rc<Summary>,
+    cur: usize,
+    dep: Rc<Dependency>,
+    prev_active: Vec<Rc<Summary>>,
+    all_candidates: Vec<Rc<Summary>>,
+}
+
+/// Recursively activates the dependencies for `top`, in depth-first order,
+/// backtracking across possible candidates for each dependency as necessary.
 ///
 /// If all dependencies can be activated and resolved to a version in the
-/// dependency graph the `finished` callback is invoked with the current state
-/// of the world.
-fn activate_deps<'a>(cx: Box<Context>,
-                     registry: &mut Registry,
-                     parent: &Summary,
-                     platform: Option<&'a str>,
-                     mut deps: slice::Iter<'a, DepInfo>,
-                     cur: usize,
-                     finished: &mut FnMut(Box<Context>, &mut Registry) -> ResolveResult)
-                     -> ResolveResult {
-    let &(dep, ref candidates, ref features) = match deps.next() {
-        Some(info) => info,
-        None => return finished(cx, registry),
-    };
+/// dependency graph, cx.resolve is returned.
+fn activate_deps_loop(mut cx: Context,
+                      registry: &mut Registry,
+                      top: Rc<Summary>,
+                      top_method: &Method) -> CargoResult<Resolve> {
+    let mut backtrack_stack: Vec<BacktrackFrame> = Vec::new();
+    let mut remaining_deps: RemainingDeps = Vec::new();
+    remaining_deps.extend(
+        try!(activate(&mut cx, registry, top, &top_method)));
```

Comment:
Stylistically Cargo tends to prefer something along the lines of:

```rust
remaining_deps.extend(try!(activate(&mut cx, registry, top,
                                    &top_method)));
```

But if this fits in 80 chars on one line then that's fine.